rlinf/hybrid_engines/sglang/common/sgl_scheduler.py

```diff
@@ -0,0 +1,225 @@
+            ]
+        )
+
+        self.is_weight_offloaded = False
+
+    def cuda_info(self, text: str = ""):
+        free_gpu_memory, total_gpu_memory = torch.cuda.mem_get_info()
+        free_gpu_memory /= 2**30
+        total_gpu_memory /= 2**30
+
+        memory_allocated = torch.cuda.memory_allocated() / 2**30
+        memory_reserved = torch.cuda.memory_reserved() / 2**30
+
+        self._rlinf_worker.log_info(
+            f"[dp {self._rlinf_worker.get_parent_rank()}-tp {self.tp_rank}] {text} "
+            f"{memory_allocated=:.2f} GiB, {memory_reserved=:.2f} GiB, "
+            f"{free_gpu_memory=:.2f} GiB, {total_gpu_memory=:.2f} GiB"
+        )
+
+    def release_memory_occupation(self, recv_req: ReleaseMemoryOccupationReqInput):
+        assert self.is_weight_offloaded is False, "Weight has been offloaded!"
+        self.is_weight_offloaded = True
+        return super().release_memory_occupation(recv_req)
+
+    def sync_hf_weight(self, recv_req: SyncHFWeightInput):
+        use_cudagraph = not self.cfg.rollout.enforce_eager
+        colocate = self.placement_mode == PlacementMode.COLLOCATED
+
+        assert use_cudagraph, "use_cudagraph must be True now."
+
+        state_dict = self._rlinf_worker.recv(
+            src_group_name=self._actor_group_name,
+            src_rank=self.actor_weight_rank,
+        )
+
+        model = self.tp_worker.worker.model_runner.model
+
+        if self.is_weight_offloaded:
+            self.resume_memory_occupation(ResumeMemoryOccupationReqInput())
+            self.is_weight_offloaded = False
+
+        if colocate:
+            for name, handle in state_dict.items():
+                func, args = handle
+                list_args = list(args)
+                # NOTE: the key is to change device id to the current device id
+                # in case two processes have different CUDA_VISIBLE_DEVICES
+                list_args[6] = torch.cuda.current_device()
+                new_weight = func(*list_args)
+
+                model.load_weights([(name, new_weight)])
+                del new_weight
+        else:
+            # disaggregate mode, recv tensor directly
+            for name, tensor in state_dict.items():
+                model.load_weights([(name, tensor)])
+        self.flush_cache()
+        return SyncHFWeightOutput()
+
+    def run_task_method(self, obj: TaskMethodInput):
+        """
+        Run a CommTask method with the given name and arguments.
+        NOTE: will call wait() if async_op is True.
+        """
+        result = getattr(self, obj.method_name)(*obj.args, **obj.kwargs)
+        if "async_op" in obj.kwargs and obj.kwargs["async_op"]:
+            result = result.wait()
+        return TaskMethodOutput(method_name=obj.method_name, result=result)
+
+    def abort_request(self, recv_req: AbortReq):
+        # Compared to the original SGLang implementation, we will remove all requests that start with the given rid.
```

Comment:
Add comment here saying that we override the abort function of sglang for compatibility with 0.4.4